### ISSUE TITLE
machined, nspawn: terminated machine entries are not removed after systemd.rpm is updated

### ISSUE THREAD

**Issue by yuwata** [systemd/systemd#7236]
### Submission type

 - Bug report

### systemd version the issue has been seen with

Recent snapshot e.g. f630daa

### Used distribution

Fedora 26 x86_64

### In case of bug report: Expected behaviour you didn't see

Terminated machines are correctly removed. New machine can successfully started.

### In case of bug report: Unexpected behaviour you saw

After updating systemd.rpm, then machined does not correctly handle machine termination.
This causes many machines in stock and new machine sometime cannot start.
```
$ systemctl -n 1000 status systemd-machined.service
● systemd-machined.service - Virtual Machine and Container Registration Service
   Loaded: loaded (/usr/lib/systemd/system/systemd-machined.service; static; vendor preset: disabled)
   Active: active (running) since Mon 2017-10-23 19:16:22 JST; 1 weeks 1 days ago
     Docs: man:systemd-machined.service(8)
           https://www.freedesktop.org/wiki/Software/systemd/machined
 Main PID: 30987 (systemd-machine)
   Status: "Processing requests..."
    Tasks: 1 (limit: 4915)
   CGroup: /system.slice/systemd-machined.service
           └─30987 /usr/lib/systemd/systemd-machined

Oct 23 19:16:22 localhost systemd[1]: Starting Virtual Machine and Container Registration Service...
Oct 23 19:16:22 localhost systemd[1]: Started Virtual Machine and Container Registration Service.
Oct 23 19:17:30 localhost systemd-machined[30987]: New machine 27340da1efa24caea3822c8c03ddc393.
Oct 23 19:17:30 localhost systemd-machined[30987]: Machine 27340da1efa24caea3822c8c03ddc393 terminated.
(snip)
Oct 28 16:37:45 localhost systemd-machined[30987]: New machine 23d5c7de50ce4c01b2939c1d6d01c181.
Oct 28 16:37:45 localhost systemd-machined[30987]: Machine 23d5c7de50ce4c01b2939c1d6d01c181 terminated.
# systemd.rpm is updated here.
Oct 30 01:59:46 localhost systemd-machined[30987]: New machine 5a7016c71b77441686550aae5f8ae671.
Oct 30 01:59:46 localhost systemd-machined[30987]: New machine c48d207e36d3461389bd3541458804a3.
Oct 30 01:59:46 localhost systemd-machined[30987]: New machine d21f40f9111e45609ff1d0a2a84c52e5.
Oct 30 01:59:46 localhost systemd-machined[30987]: New machine 6731ae194d8541268b1ce270ce831c23.
(snip)
Oct 31 21:02:28 localhost systemd-machined[30987]: New machine fd06dd61939149ca9897d7c9aae2a5e7.
Oct 31 21:02:37 localhost systemd-machined[30987]: New machine 94a4995fa07b4ec1b9577fefc5a4a63c.
```
```
$ machinectl
MACHINE                          CLASS     SERVICE        OS VERSION ADDRESSES
03740dd800f04a11bad533a6cc7660b2 container systemd-nspawn -  -       -
(snip)
ff6380511fd349cca0bac8897b4b94b5 container systemd-nspawn -  -       -

90 machines listed.

```
I mainly use nspawn by [mock](https://github.com/rpm-software-management/mock/). After systemd.rpm is updated, mock logs the following WARNING:
```
WARNING: Machine 37468924c3d44062866694f1ecb325bc still running. Killing...
```
And sometimes mock fails to start new machine and shows the following ERROR. In the build.log,
```
ENTER ['do'](['bash', '--login', '-c', '/usr/bin/rpmbuild -bb --target x86_64 --nodeps /builddir/build/SPECS/systemd.spec'], chrootPath='/var/lib/mock/fedora-27-x86_64-systemd-1130/root'env={'TERM': 'vt100', 'SHELL': '/bin/bash', 'HOME': '/builddir', 'HOSTNAME': 'mock', 'PATH': '/usr/bin:/bin:/usr/sbin:/sbin', 'PROMPT_COMMAND': 'printf "\\033]0;<mock-chroot>\\007"', 'PS1': '<mock-chroot> \\s-\\v\\$ ', 'LANG': 'en_US.utf8'}shell=Falselogger=<mockbuild.trace_decorator.getLog object at 0x7f352bf711d0>timeout=0uid=1014gid=135user='mockbuild'nspawn_args=['--private-network']printOutput=False)
Executing command: ['/usr/bin/systemd-nspawn', '-q', '-M', '925b000c1bfa4645bfd0d18075f530f8', '-D', '/var/lib/mock/fedora-27-x86_64-systemd-1130/root', '-a', '--private-network', '--setenv=TERM=vt100', '--setenv=SHELL=/bin/bash', '--setenv=HOME=/builddir', '--setenv=HOSTNAME=mock', '--setenv=PATH=/usr/bin:/bin:/usr/sbin:/sbin', '--setenv=PROMPT_COMMAND=printf "\\033]0;<mock-chroot>\\007"', '--setenv=PS1=<mock-chroot> \\s-\\v\\$ ', '--setenv=LANG=en_US.utf8', '-u', 'mockbuild', 'bash', '--login', '-c', '/usr/bin/rpmbuild -bb --target x86_64 --nodeps /builddir/build/SPECS/systemd.spec'] with env {'TERM': 'vt100', 'SHELL': '/bin/bash', 'HOME': '/builddir', 'HOSTNAME': 'mock', 'PATH': '/usr/bin:/bin:/usr/sbin:/sbin', 'PROMPT_COMMAND': 'printf "\\033]0;<mock-chroot>\\007"', 'PS1': '<mock-chroot> \\s-\\v\\$ ', 'LANG': 'en_US.utf8'} and shell False
Failed to register machine: File exists
Child return code was: 1
EXCEPTION: [Error()]
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/mockbuild/trace_decorator.py", line 96, in trace
    result = func(*args, **kw)
  File "/usr/lib/python3.6/site-packages/mockbuild/util.py", line 598, in do
    raise exception.Error("Command failed: \n # %s\n%s" % (command, output), child.returncode)
mockbuild.exception.Error: Command failed: 
 # /usr/bin/systemd-nspawn -q -M 925b000c1bfa4645bfd0d18075f530f8 -D /var/lib/mock/fedora-27-x86_64-systemd-1130/root -a --private-network --setenv=TERM=vt100 --setenv=SHELL=/bin/bash --setenv=HOME=/builddir --setenv=HOSTNAME=mock --setenv=PATH=/usr/bin:/bin:/usr/sbin:/sbin --setenv=PROMPT_COMMAND=printf "\033]0;<mock-chroot>\007" --setenv=PS1=<mock-chroot> \s-\v\$  --setenv=LANG=en_US.utf8 -u mockbuild bash --login -c /usr/bin/rpmbuild -bb --target x86_64 --nodeps /builddir/build/SPECS/systemd.spec
```
### In case of bug report: Steps to reproduce the problem

Update systemd.rpm, then start many machines.

**Comment by yuwata** [comment]
After restarting systemd-machined, terminated machines are cleared and new machines can start correctly.

**Comment by tomhughes** [comment]
It looks like systemd-machined never gets the `MachineRemoved` signal on the bus - in [dbus.zip](https://github.com/systemd/systemd/files/1646385/dbus.zip) are two busctl pcap files, `good.pcap` before systemd is restarted and `bad.pcap` afterwards.

In each case I started the trace then started a mock shell and exited it before stopping the trace. Bother traces show two `MachineNew` signals but only the first has the matching `MachineRemoved` signals.

**Comment by tomhughes** [comment]
Oh hang on those are outgoing signals so that's expected if it hasn't realised the machine has gone.

**Comment by tomhughes** [comment]
After more investigation the problem seems to be that `systemd` is not sending `UnitRemoved` signals on the bus that `systemd-machined` uses to know when machines go away.

I can see that `systemd` only seems to send those on the API bus if it thinks there is somebody subscribed, and that `systemd-machined` subscribes at startup but doesn't do anything about resubscribing if `systemd` restarts.

Equally though it looks like `systemd` tries to serialize the list of subscribers and restore it when restarting.

**Comment by keszybz** [comment]
Pfff, I tried to debug this. I had something like ~150 dead machines listed by machinectl, mostly from old mock runs. I thought I'd "clear the slate", so I restarted everything, and now for I can't reproduce the bug any more. I tried all combinations of restarting machined, systemd, and the machines themselves, and everything works.

The latest commits that could be related seems to be e63ebf71edd7947f29389c72e851d8df5c7bedda are 31dc1ca3bf9fbb0baf308d213e58ea389cceff4e.
